pkg/universe.dagger.io/CODING_STYLE.md

```diff
@@ -0,0 +1,203 @@
+
+// bad, starts with lower case but not private
+#mergeStructs: { ... }
+
+// bad, starts with upper case but private
+_#MergeStructs: { ... }
+```
+
+Choose `_camelCase` for private fields (implementation details).
+
+```cue
+#Copy: {
+    // good, indicates internal implementation
+    _copy: core.#Copy
+
+    // bad, indicates it can be used directly
+    copy: core.#Copy
+}
+```
+
+## Definitions (`#`) for *schemas*, fields for concrete *implementations*
+
+```cue
+// good, defines a schema
+#Copy: {
+    // good, implements a schema
+    input: #Image
+    _copy: core.#Copy & { ... }
+}
+
+// bad, not a definition
+Copy: {
+    // bad, not using a visible or hidden field (according to intent)
+    #input: #Image
+    _#copy: core.#Copy & { ... }
+}
+
+// ok for mixin
+Mixin: { ... }
+copy: #Copy & Mixin & { ... }
+```
+
+## All public definitions and fields must be documented
+
+Other fields are optional. The main idea is that public fields should be documented for someone
+reading Dagger’s documentation in the browser. Everything else is for someone looking at the code.
+
+```yaml
+// Copy files into an image
+#Copy: {
+    // Image to receive copied files
+    input: #Image
+
+    // Files to copy
+    contents: dagger.#FS
+
+    _copy: ...
+
+    // Resulting image with added files
+    output: ...
+}
+```
+
+## Use verbs for action names, nouns otherwise
+
+Action definition names should be verbs (e.g., `#Build`, `#Run`, `#Push`),
+otherwise they should be nouns as much as possible (e.g., `#Image`, `#Secret`, `#Socket`).
+
+## No field abbreviations
+
+CUE fields are more explicit and intuitive without abbreviations (e.g, `target` or
+`destination` instead of `dest`).
+
+## Don’t inline scripts
+
+Avoid inlining scripts (e.g., *sh*, *py*, etc). Instead put them in their own files,
+with proper extension, and use `core.#Source` to import into CUE. This allows linting
+and avoids some limitations (script size, escaping).
+
+Some exceptions may apply:
+
+- The script is really short;
+- Interpolation is necessary (although there may be other solutions).
+
+```cue
+// good
+src: core.#Source & {
+    path: "./src"
+}
+run: bash.#Run & {
+    script: {
+        directory: src.output
+        filename: "setup.sh"
+    }
+}
+
+// ok
+run: bash.#Run & {
+    script: contents: "echo hello world"
+}
+
+// bad, better solution available
+let name = "world"
+run: bash.#Run & {
+    script: contents: """
+        ...
+        echo hello \(world)
+        ...
+        """
+}
+
+// good, use env var in script
+run: bash.#Run & {
+    env: NAME: name
+    script: filename: "setup.sh"
+}
+```
+
+### One definition per file
+
+Split action definitions into their own file as much as possible (unless very tightly related).
+
+```yaml
+// copy.cue
+#Copy: ...
+
+// push.cue
+#Push: ...
+
+// exec.cue
+#Exec: ...
+#Mount: ...
+#CacheDir: ...
+```
+
+#### Avoid `#”””`
+
+Don’t use `#”””` unless you actually need it. Most of the time you won't.
```

Comment:
Maybe explain when we should  need it?

Reply:
I tried to find an example where we would need it. If you know of one, let me know.